docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,236 @@
+      # name: my-vnet
+        cidr: 10.250.0.0/16
+      workers: 10.250.0.0/19
+status:
+  state: |
+      {
+          "version": 3,
+          "terraform_version": "0.11.14",
+          "serial": 2,
+          "lineage": "3a1e2faa-e7b6-f5f0-5043-368dd8ea6c10",
+          "modules": [
+              {
+              }
+          ]
+          ...
+      }
+```
+
+Extensions which do not require state migration should set `status.state=nil` in their Custom Resources and trigger a normal reconciliation operation if the CR contains the `core.gardener.cloud/operation=restore` annotation.
+
+**Note:** The same size limitations from the previous section are relevant here as well.
+
+### Shoot reconciliation flow changes
+
+The only data which must be stored in the `ShootState` by the Garden Controller Manager is secrets (e.g ca for the apiserver). Therefore the `botanist.DeploySecrets` step is changed. It is split into two functions which take a list of secrets that have to be generated.
+- `botanist.GenerateSecretState` Generates certificate authorities and other secrets which have to be persisted in the ShootState and must not be regenerated on the New Seed.
+- `botanist.DeploySecrets` Takes secret data from the `ShootState`, generates new ones (e.g. client tls certificates from the saved certificate authorities) and deploys everything in the Shoot's control plane on the New Seed
+
+
+### ShootState synchronization controller
+
+The ShootState synchronization controller is part of the garden controller manager. It syncs the state of extension custom resources from the shoot namespace to the garden cluster and updates the corresponding `spec.extension.state` field in the `ShootState` resource. The controller can `watch` Custom Resources used by the extensions and update the `ShootState` only when changes occur.
+
+
+### Migration workflow
+1. Starting migration
+    - Migration can only be started after a Shoot cluster has been successfully created so that the `status.seed` field in the `Shoot` resource has been set
+    - The `Shoot` resource's field `spec.seedName="new-seed"` is edited to hold the name of the __Destination Seed__ and reconciliation is automatically triggered
+    - The Garden Controller Manager checks if the equality between `spec.seedName` and `status.seed`, detects that they are different and triggers migration.
+1. The Garden Controller Manager waits for the __Destination Seed__ to be ready
+    - The Garden Controller Manager lists all deployed extension resources for the Shoot and copies the values in their `status.state` field to the `ShootState` resource
+1. The Shoot State synchronization controller in GCM syncs the `ShootState` from the __Source Seed__ cluster. Later this will be handled by gardenlet (This might not be necessary when using `watches`).
+    - If the __Source Seed's__ API server is not available, then the last successfully synced `ShotoState` is used
+1. Stop the Shoot's API server
```

Comment:
Not sure if this can be handled by simply first deleting the finalizers and then deleting the resources. 

But still you would have to somehow tell the worker controller to remove the finalizers on the MCM objects before deleting them.

Reply:
Do you have concrete examples for what you mean with "migration-related cleanup"?
Ideally, there is nothing that needs to be done BEFORE we move the control plane to a new seed. Only after the migration is completed we should instruct the extensions to do "migration-related cleanup", as you say, without actually deleting infrastructural resources.

Reply:
@rfranzke I talked with @swapnilgm. Here's an example:
> An example of migration related cleanup would be removing MCM-related CRDs and secrets (objects with the MCM finalizer) without actually deleting machines. MCM would be good to do the cleanup as it only knows which objects have that finalizer.  So gardenlet would have to tell the worker controller and it would then have to tell MCM to do this cleanup.
> If we remove finalizers externally there could be a chance for a race condition if the reconciliation annotation was set on an extension resource and it is queued for reconciliation.


Generally we want to avoid having an extension resource queued for reconciliation once migration has started as that could cause other problems (split-brain scenario) and the MCM controller should have been stopped at the point where resource cleanup is happening.

Reply:
> Generally we want to avoid having an extension resource queued for reconciliation once migration has started 

OK, makes sense. However, we start with only allowing changing the seed name if the shoot is reconciled successfully in the beginning, but eventually, we want to also allow it if the reconciliation fails (which it will in case the seed is broken). So, how do we ensure this then? Maybe we can't talk to the kube-apiserver of the old seed anymore (in which case the extensions can't do anything either), but what if it comes up again and we are in the middle of the migration?

> and the MCM controller should have been stopped at the point where resource cleanup is happening.

Yes, for the "happy case" we can simply scale down everything, but what if the old seed is not reachable anymore?

Reply:
> and the MCM controller should have been stopped at the point where resource cleanup is happening.

let me correct on this, rather than bringing MCM controller down, only point is the MCM controller to not execute reconciliation or actual deletion code path. i.e. MCM controller can still be up and running but looking at MCM object in source seed once the migration is started 

> but what if it comes up again and we are in the middle of the migration?.
> but what if the old seed is not reachable anymore?
The controlplane-coordinator running as a part of gardenelet or some dependency watchdog based on our decision with leader election will take over in that case. The annotaion business from gardener is just to optimize time on those cleanup. Cause it knows the exact time when the migration was requested and if source seed apiserver is available, it can start instruct the migration related activity there

Reply:
Indeed if the reconciliation fails and there are reachability problems between the source seed and the garden cluster we can't know if there are custom resourceswhich were already being reconciled and failed and get constantly requeued. 
Then we need to rely on an annotation or a leader election functionality in the extension controllers to make them stop reconciliation.

Reply:
@rfranzke @swapnilgm wdyt of proceeding like this for now:
Gardenlet adds the "migration" annotation (we can rename it) to extension resources so that they can be safely removed if the extension controllers are constantly requeueing them for reconciliation due to errors.

We can think of what to do in the case when there are problems with the seed's apiserver and no reachability between seed apiserver and garden cluster once we come up with a clear strategy for leader election.

However, I am still wondering whether we should leave the cleanup to the extension controllers or to gardenlet. The latter would require less code so it would be easier to change later on  (if necessary) when we also implement leader election.

Reply:
Agreed. It makes more sense to shape the details once we have agreed on a general leader election mechanism.
Regarding the cleanup, I think the only choice is to make the extension controllers responsible. Only they know which resources they created and what is the best way to clean them up.

Reply:
Same here. Agreed.